File: backend/app/api/routes/audit_logs.py
import math

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import require_admin
from app.core.database import get_db
from app.crud import audit as audit_crud
from app.models.user import User
from app.schemas.audit_log import AuditLogOut
from app.schemas.common import Page

router = APIRouter(prefix="/audit-logs", tags=["audit-logs"])


@router.get("", response_model=Page[AuditLogOut])
def list_audit_logs(
    page: int = 1,
    page_size: int = 50,
    entity_type: str | None = None,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
) -> Page[AuditLogOut]:
    items, total = audit_crud.list_paginated(db, page, page_size, entity_type)
    out = [
        AuditLogOut(
            id=log.id,
            user_id=log.user_id,
            username=log.user.username if log.user else None,
            action=log.action,
            entity_type=log.entity_type,
            entity_id=log.entity_id,
            details=log.details,
            created_at=log.created_at,
        )
        for log in items
    ]
    return Page(items=out, total=total, page=page, page_size=page_size, total_pages=max(math.ceil(total / page_size), 1))

File: workers/src/routes/audit_logs.ts
import { Hono } from "hono";

import { intQueryParam } from "../lib/validate";
import { requireAdmin, requireUser, type AppEnv } from "../middleware/auth";
import * as auditRepo from "../repo/audit";

export const auditLogRoutes = new Hono<AppEnv>();

auditLogRoutes.use("*", requireUser, requireAdmin);

auditLogRoutes.get("/", async (c) => {
  const page = intQueryParam("page", c.req.query("page"), 1);
  const pageSize = intQueryParam("page_size", c.req.query("page_size"), 50);
  const entityType = c.req.query("entity_type");

  const { items, total } = await auditRepo.listPaginated(c.env.DB, page, pageSize, entityType);

  return c.json({
    items: items.map((log) => ({
      id: log.id,
      user_id: log.user_id,
      email: log.email,
      action: log.action,
      entity_type: log.entity_type,
      entity_id: log.entity_id,
      // Stored as TEXT; the API contract is a JSON object, as the SQLAlchemy JSON column was.
      details: safeParse(log.details),
      created_at: log.created_at,
    })),
    total,
    page,
    page_size: pageSize,
    total_pages: Math.max(Math.ceil(total / pageSize), 1),
  });
});

function safeParse(raw: string | null): Record<string, unknown> {
  if (!raw) return {};
  try {
    const parsed: unknown = JSON.parse(raw);
    return parsed && typeof parsed === "object" ? (parsed as Record<string, unknown>) : {};
  } catch {
    return {};
  }
}
